ios/AppLovinMAXAdView.m

```diff
@@ -0,0 +1,189 @@
+@property (nonatomic, copy, nullable) NSString *placement;
+@property (nonatomic, copy, nullable) NSString *customData;
+@property (nonatomic, assign, readonly, getter=isAdaptiveBannerEnabled) BOOL adaptiveBannerEnabled;
+@property (nonatomic, assign, readonly, getter=isAutoRefresh) BOOL autoRefresh;
+
+@end
+
+@implementation AppLovinMAXAdView
+
+- (void)setAdUnitId:(NSString *)adUnitId
+{
+    // Ad Unit ID must be set prior to creating MAAdView
+    if ( self.adView )
+    {
+        [[AppLovinMAX shared] log: @"Attempting to set Ad Unit ID %@ after MAAdView is created", adUnitId];
+        return;
+    }
+    
+    _adUnitId = adUnitId;
+    
+    [self attachAdView];
+}  
+
+- (void)setAdFormat:(NSString *)adFormat
+{
+    // Ad format must be set prior to creating MAAdView
+    if ( self.adView )
+    {
+        [[AppLovinMAX shared] log: @"Attempting to set ad format %@ after MAAdView is created", adFormat];
+        return;
+    }
+    
+    if ( [@"banner" isEqualToString: adFormat] )
+    {
+        _adFormat = DEVICE_SPECIFIC_ADVIEW_AD_FORMAT;
+    }
+    else if ( [@"mrec" isEqualToString: adFormat] )
+    {
+        _adFormat = MAAdFormat.mrec;
+    }
+    else
+    {
+        [[AppLovinMAX shared] log: @"Attempting to set an invalid ad format of \"%@\" for %@", adFormat, self.adUnitId];
+        return;
+    }
+    
+    [self attachAdView];
+}  
+
+- (void)setPlacement:(NSString *)placement
+{
+    _placement = placement;
+    
+    if ( self.adView )
+    {
+        self.adView.placement = placement;
+    }
+}
+
+- (void)setCustomData:(NSString *)customData
+{
+    _customData = customData;
+    
+    if ( self.adView )
+    {
+        self.adView.customData = customData;
+    }
+}
+
+- (void)setAdaptiveBannerEnabled:(BOOL)adaptiveBannerEnabled
+{
+    _adaptiveBannerEnabled = adaptiveBannerEnabled;
+    
+    if ( self.adView )
+    {
+        [self.adView setExtraParameterForKey: @"adaptive_banner" value: adaptiveBannerEnabled ? @"true" : @"false"];
+    }
+}
+
+- (void)setAutoRefresh:(BOOL)autoRefresh
+{
+    _autoRefresh = autoRefresh;
+    
+    if ( self.adView )
+    {
+        if ( autoRefresh )
+        {
+            [self.adView startAutoRefresh];
+        }
+        else
+        {
+            [self.adView stopAutoRefresh];
+        }
+    }
+}
+
+- (void)attachAdView
+{
+    if ( ![self.adUnitId al_isValidString] )
+    {
+        [[AppLovinMAX shared] log: @"Attempting to attach MAAdView without Ad Unit ID"];
+        return;
+    }
+    
+    // Re-assign in case of race condition
+    NSString *adUnitId = self.adUnitId;
+    MAAdFormat *adFormat = self.adFormat;
+    
+    // Run after 0.25 sec delay to allow all properties to set
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) (NSEC_PER_SEC/4)), dispatch_get_main_queue(), ^{
+        if ( !self.adFormat )
+        {
+            [[AppLovinMAX shared] log: @"Attempting to attach MAAdView without ad format"];
+            return;
+        }
```

Comment:
Why's the ad format check in the block, but the `if ( ![self.adUnitId al_isValidString] )` check not? They should be right next to each other.

Reply:
Moved inside the runnable on both platforms.